app/models/concerns/with_assignments_batch.rb

```diff
@@ -0,0 +1,31 @@
+# WithAssignmentsBatch mirrors the WithAssignment mixin
+# but implements operations in batches, so that they outperform
+# their counterparts
+module WithAssignmentsBatch
+  extend ActiveSupport::Concern
+
+  def find_assignments_for(user, _organization = Organization.current, &block)
+    block = block_given? ? block : lambda { |it, _e| it }
+
+    return exercises.map { |it| block.call nil, it  } unless user
+
+    pairs = exercises.map { |it| [it.id, [nil, it]] }.to_h
+    Assignment.where(submitter: user, exercise: exercises).each do |it|
+      pairs[it.exercise_id][0] = it
+    end
+
+    pairs.values.map { |assignment, exercise| block.call assignment, exercise }
+  end
```

Comment:
I find having both `assignments_for` and `find_assignments_for` a tad confusing. Does `find_assignment_for` have any purpose other than being of use to `statuses_for` and `assignments_for`? Are there any use cases where we would want to call it directly?

I thought of suggesting it should be private but I see you're using it to test on `guide_spec.rb`.

Reply:
I see your point. I simply put it public to mimic the `WithAssignments` mixin, which also has a `find_assignment_for` with the same semantics: _find_ it returning nil if it does not exist, in the same manner as the `find_by` methods. 

It may also be useful for other implementations, but it is mostly for sake of consistency. Both methods may turn private in the future.